staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/metrics.go

```diff
@@ -0,0 +1,119 @@
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"k8s.io/utils/clock"
+	"sync"
+	"time"
+
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "authentication"
+)
+
+var (
+	jwtAuthenticatorLatencyMetric = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "jwt_authenticator_latency_seconds",
+			Help:           "Latency of jwt authentication operations in seconds. This is the time spent authenticating a token for cache miss only (i.e. when the token is not found in the cache).",
+			StabilityLevel: metrics.ALPHA,
+			// Start 0.1ms with the last bucket being [~52s, +Inf)
+			Buckets: metrics.ExponentialBuckets(0.0001, 2, 20),
+		},
+		[]string{"result", "jwt_issuer_hash", "apiserver_id_hash"},
+	)
+)
+
+var registerMetrics sync.Once
+var hashPool *sync.Pool
+
+func RegisterMetrics() {
+	registerMetrics.Do(func() {
+		hashPool = &sync.Pool{
+			New: func() interface{} {
+				return sha256.New()
+			},
+		}
+		legacyregistry.MustRegister(jwtAuthenticatorLatencyMetric)
+	})
+}
+
+func recordAuthenticationLatency(result, jwtIssuerHash, apiServerIDHash string, duration time.Duration) {
+	jwtAuthenticatorLatencyMetric.WithLabelValues(result, jwtIssuerHash, apiServerIDHash).Observe(duration.Seconds())
+}
+
+func getHash(data string) string {
+	if len(data) == 0 {
+		return ""
+	}
+	h := hashPool.Get().(hash.Hash)
+	h.Reset()
+	h.Write([]byte(data))
+	dataHash := fmt.Sprintf("sha256:%x", h.Sum(nil))
+	hashPool.Put(h)
+	return dataHash
+}
+
+func InstrumentedAuthenticator(jwtIssuer, apiServerID string, delegate authenticator.Token) authenticator.Token {
+	return newInstrumentedAuthenticator(jwtIssuer, apiServerID, delegate, clock.RealClock{})
+}
+
+func newInstrumentedAuthenticator(jwtIssuer, apiServerID string, delegate authenticator.Token, clock clock.PassiveClock) *instrumentedAuthenticator {
+	return &instrumentedAuthenticator{
+		jwtIssuerHash:   getHash(jwtIssuer),
+		apiServerIDHash: getHash(apiServerID),
+		delegate:        delegate,
+		clock:           clock,
+	}
+}
+
+type instrumentedAuthenticator struct {
+	jwtIssuerHash   string
+	apiServerIDHash string
+	delegate        authenticator.Token
+	clock           clock.PassiveClock
+}
+
+func (a *instrumentedAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+	start := a.clock.Now()
+	response, ok, err := a.delegate.AuthenticateToken(ctx, token)
+	// this only happens when issuer doesn't match the authenticator
+	// we don't want to record metrics for this case
+	if !ok && err == nil {
+		return response, ok, err
+	}
+
+	duration := a.clock.Since(start)
+	if err != nil {
+		recordAuthenticationLatency("failed", a.jwtIssuerHash, a.apiServerIDHash, duration)
```

Comment:
other places we use `failure / success`, not `failed / success` ... should we be consistent?